### PR TITLE
[#151] Persistent DB

### LINKED ISSUE #151: A persistent DB
Indexing is an expensive operation. We probably want to persist the indexed data on disk across sessions.

### PR DESCRIPTION
**Pull request by robertoaloi**
### Description

A port of the current solution to Mnesia. Sorry about the huge size, but it wasn't something easy to do in small chunks.
With this PR, a database instance is created per project/otp version.
The aim of this PR is to gain persistency whilst not introducing any regression.

Closes #151 .

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -16,10 +16,12 @@
        , {docsh,       "0.7.2"}
        , {worker_pool, "4.0.1"}
        , {elvis,       "0.4.3", {pkg, elvis_core}}
+       , {uuid,        "1.8.0", {pkg, uuid_erl}}
```
Comment (robertoaloi) [inline rebar.config]: This has been included to produce unique ids, not used with the current schema.

File: rebar.config [hunk 2]
```diff
@@ -90,3 +92,5 @@
 %% Disable warning_as_errors for redbug to avoid deprecation warnings.
 {overrides, [ {del, redbug, [{erl_opts, [warnings_as_errors]}]}
             ]}.
+
+{dist_node, [ {sname, 'erlang_ls'} ]}.
```
Comment (robertoaloi) [inline rebar.config]: Most rebar3 commands (e.g. `ct` will use this to start the node in distributed mode).

File: src/els_completion_provider.erl [hunk 3]
```diff
@@ -258,6 +260,7 @@ keywords() ->
 %% Filter by prefix
 %%==============================================================================
 
+%% TODO: Implement as select
```
Comment (robertoaloi) [inline src/els_completion_provider.erl]: Most of the filtering/selecting in the providers can be implemented as Mnesia queries.

File: src/els_db.erl [hunk 4]
```diff
@@ -1,133 +1,93 @@
...
+  create_tables(),
+  application:stop(mnesia).
+
+-spec create_tables() -> ok.
+create_tables() ->
+  [els_db_table:create(T) || T <- ?TABLES],
   ok.
 
--spec update(table(), key(), function(), any()) -> ok.
-update(Table, Key, UpdateFun, Default) ->
```
Comment (robertoaloi) [inline src/els_db.erl]: It turns out this function was unused?

File: src/els_db.erl [hunk 5]
```diff
@@ -1,133 +1,93 @@
...
-  end.
-
--spec delete(table(), key()) -> ok.
+-spec delete(atom(), any()) -> ok | {error, any()}.
 delete(Table, Key) ->
-  true = ets:delete(Table, Key),
+  F = fun() -> mnesia:delete({Table, Key}) end,
+  transaction(F).
+
+-spec lookup(atom(), any()) -> {ok, [tuple()]} | {error, any()}.
```
Comment (robertoaloi) [inline src/els_db.erl]: We probably don't need to run everything as a transaction.

File: src/els_document.erl [hunk 6]
```diff
@@ -11,6 +11,7 @@
 -export([ create/2
         , uri/1
         , text/1
+        , md5/1
```
Comment (robertoaloi) [inline src/els_document.erl]: Used to check whether the version of the document is already indexed.

File: src/els_utils.erl [hunk 7]
```diff
@@ -41,13 +41,14 @@ find_module(M, Extension) ->
...
   {ok, els_document:document()} | {error, any()}.
 find_document(Uri) ->
-  case els_db:find(documents, Uri) of
-    {ok, Document} ->
+  case els_dt_documents:lookup(Uri) of
+    {ok, [#{document := Document}]} ->
       {ok, Document};
-    {error, not_found} ->
+    {ok, []} ->
+      %% TODO: Maybe we should not index this
```
Comment (robertoaloi) [inline src/els_utils.erl]: I am often getting timeouts when using this branch and navigating towards a OTP module during indexing.

File: src/erlang_ls.erl [hunk 8]
```diff
@@ -12,6 +12,7 @@ main(Args) ->
   ok = parse_args(Args),
   ok = init_node_name(is_debug()),
   ok = lager_config(),
+  els_db:install(),
```
Comment (robertoaloi) [inline src/erlang_ls.erl]: The DB is currently recreated on startup. Obviously this needs to be fixed.

File: src/erlang_ls.app.src [hunk 9]
```diff
@@ -29,6 +30,7 @@
         %% Logging
       , {logging_enabled, false}
       , {log_dir, "/usr/local/var/log/erlang_ls"}
+      , {db_dir, "/usr/local/var/cache/erlang_ls"}
```
Comment (alanz) [inline src/erlang_ls.app.src]: Will this work on different operating systems?

Comment (robertoaloi) [reply src/erlang_ls.app.src]: Fair point. The logging would also break, but that's disabled by default.
On the other hand, if we aim at supporting multiple platforms we should somehow add them to the CI pipelines.

File: rebar.config (hunk 2 above)
Comment (jfacorro) [reply rebar.config]: Do we need this though?

File: src/els_db.erl (hunk 4 above)
Comment (jfacorro) [reply src/els_db.erl]: We stopped using it when we changes the `references` table to a `bag`.

File: src/els_dt_documents.erl [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+%%==============================================================================
+%% Item Definition
+%%==============================================================================
+
+-record(els_dt_documents, { uri      :: uri()  | '_'
+                          , document :: els_document:document()
+                          }).
+-type els_dt_documents() :: #els_dt_documents{}.
+
+-type item() :: #{ uri      := uri()
```
Comment (jfacorro) [inline src/els_dt_documents.erl]: The document already contains the `uri`, why not just have the following?
```
-type item() :: els_document:document().
```

File: src/els_dt_documents.erl [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+%%==============================================================================
+%% API
+%%==============================================================================
+
+-spec from_map(item()) -> els_dt_documents().
+from_map(#{ uri := Uri, document := Document }) ->
+  #els_dt_documents{ uri = Uri, document = Document }.
+
+-spec to_map(els_dt_documents()) -> item().
+to_map(#els_dt_documents{ uri = Uri, document = Document }) ->
```
Comment (jfacorro) [inline src/els_dt_documents.erl]: This could be renamed to `to_item` if the suggestion above is applied.

File: src/els_dt_documents.erl [hunk 12]
```diff
@@ -0,0 +1,83 @@
+%%==============================================================================
+%% The 'documents' table
+%%==============================================================================
+
+-module(els_dt_documents).
+
```
Comment (jfacorro) [inline src/els_dt_documents.erl]: Isn't the `-behaviour(els_db_table).` missing from all `*_dt_*` modules?

File: src/els_document_highlight_provider.erl [hunk 13]
```diff
@@ -52,21 +52,22 @@ find_highlights(Uri, #{ kind := Kind
...
+          ],
       R
   end;
 find_highlights(_Uri, _POI) ->
   null.
 
 -spec document_highlight(poi_range()) -> map().
 document_highlight(Range) ->
-    %% TODO:AZ:range as displayed ends one col too early
+  %% TODO:AZ:range as displayed ends one col too early
```
Comment (jfacorro) [inline src/els_document_highlight_provider.erl]: This was probably fixed already by [this change](https://github.com/erlang-ls/erlang_ls/pull/280/files#diff-3f144025dae38c6a860fe792f9f86294R26).

File: src/els_dt_documents.erl [hunk 14]
```diff
@@ -0,0 +1,83 @@
+%%==============================================================================
+%% The 'documents' table
+%%==============================================================================
+
+-module(els_dt_documents).
```
Comment (jfacorro) [inline src/els_dt_documents.erl]: What does `dt` stand for here?

File: src/erlang_ls.erl (hunk 8 above)
Comment (jfacorro) [reply src/erlang_ls.erl]: What happens in debug mode when the node's name is always different? Have you tried this branch out with different `erlang_ls` instances working at the same time?

File: src/erlang_ls.app.src (hunk 9 above)
Comment (michalmuskala) [reply src/erlang_ls.app.src]: [`filename:basedir/3`](http://erlang.org/doc/man/filename.html#basedir-3) might be useful to get correct temp or config directories on various systems

File: src/els_dt_documents.erl (hunk 10 above)
Comment (robertoaloi) [reply src/els_dt_documents.erl]: That's exactly where I'm heading to in the follow-up commits (not pushed yet).

File: src/els_code_navigation.erl [hunk 15]
```diff
@@ -51,9 +51,8 @@ goto_definition(Uri, #{ kind := record_expr, id := Record }) ->
 goto_definition(_Uri, #{ kind := Kind, id := Include }
                ) when Kind =:= include;
                       Kind =:= include_lib ->
-  FileName = filename:basename(Include),
-  M = list_to_atom(FileName),
-  case els_utils:find_module(M) of
+  H = list_to_atom(filename:basename(Include, filename:extension(Include))),
```
Comment (robertoaloi) [inline src/els_code_navigation.erl]: The "module" table is now gone, in favour of an indexed attribute on the "document" table. Now, header files are stored as "documents" as well and their id does not include the 'hrl' extension anymore, hence this change.

File: src/els_code_navigation.erl [hunk 16]
```diff
@@ -51,9 +51,8 @@ goto_definition(Uri, #{ kind := record_expr, id := Record }) ->
...
+  case els_utils:find_header(H) of
```
Comment (robertoaloi) [inline src/els_code_navigation.erl]: This is something that annoyed me for a long time. We were looking for modules named 'xxx.hrl'.

File: src/els_code_navigation.erl [hunk 17]
```diff
@@ -72,38 +71,34 @@ goto_definition(_Filename, _) ->
 find([], _Kind, _Data) ->
   {error, not_found};
 find([Uri|Uris0], Kind, Data) ->
-  case els_db:find(documents, Uri) of
-    {ok, Document} ->
-      POIs = els_document:points_of_interest(Document, [Kind], Data),
-      case POIs of
+  case els_dt_document:lookup(Uri) of
```
Comment (robertoaloi) [inline src/els_code_navigation.erl]: To support both sets and bags, the lookup API currently returns a list. We may want to change this later.

File: src/els_document_highlight_provider.erl (hunk 13 above)
Comment (robertoaloi) [reply src/els_document_highlight_provider.erl]: Removed.

File: src/erlang_ls.app.src (hunk 9 above)
Comment (robertoaloi) [reply src/erlang_ls.app.src]: Nice! I didn't know about it.

Comment (robertoaloi) [reply src/erlang_ls.app.src]: That's what happens when you tend to only support one system :p

File: src/erlang_ls.erl (hunk 8 above)
Comment (robertoaloi) [reply src/erlang_ls.erl]: Things will probably explode. That's one of the reasons why this branch is not ready for merge :)

File: test/els_completion_SUITE.erl [hunk 18]
```diff
@@ -88,10 +88,6 @@ atom_completions(Config) ->
                 , kind             => ?COMPLETION_ITEM_KIND_MODULE
                 , label            => <<"code_navigation_types">>
                 }
-             , #{ insertTextFormat => ?INSERT_TEXT_FORMAT_PLAIN_TEXT
```
Comment (robertoaloi) [inline test/els_completion_SUITE.erl]: I think this was an actual bug. We were including header files as completions for atoms. This was due to the fact we were storing modules and headers together. With the new implementation this is not the case anymore.

File: test/els_completion_SUITE.erl [hunk 19]
```diff
@@ -270,11 +266,11 @@ macros(Config) ->
 
   #{result := Completion1} =
     els_client:completion(Uri, 24, 1, TriggerKindChar, <<"?">>),
-  ?assertEqual(lists:sort(Expected), lists:sort(Completion1)),
+  [?assert(lists:member(E, Completion1)) || E <- Expected],
```
Comment (robertoaloi) [inline test/els_completion_SUITE.erl]: Another spotted bug: for some reason, macros from included files were not returned in the completion list. This is now fixed.

File: rebar.config (hunk 1 above)
Comment (robertoaloi) [reply rebar.config]: Ended up removing this, since not in scope.

File: rebar.config (hunk 2 above)
Comment (robertoaloi) [reply rebar.config]: I don't recall why I had to add it. It was related to the Mnesia setup. Will check...

File: src/els_dt_documents.erl (hunk 10 above)
Comment (robertoaloi) [reply src/els_dt_documents.erl]: I ended up merging the `els_document` and `els_dt_document` entities into one. I also included the `kind` in the document and got rid of the `modules` table.

File: src/els_dt_documents.erl (hunk 11 above)
Comment (robertoaloi) [reply src/els_dt_documents.erl]: Sounds reasonable. Will do.

File: src/els_dt_documents.erl (hunk 12 above)
Comment (robertoaloi) [reply src/els_dt_documents.erl]: 🤦‍♂

File: src/els_dt_documents.erl (hunk 14 above)
Comment (robertoaloi) [reply src/els_dt_documents.erl]: Db Table. It's the name of the behaviour. I'm open for suggestion.

File: src/els_utils.erl [hunk 20]
```diff
@@ -1,12 +1,11 @@
 -module(els_utils).
 
--export([ find_document/1
+-export([ find_header/1
         , find_module/1
-        , find_module/2
         , fold_files/4
+        , lookup_document/1
```
Comment (jfacorro) [inline src/els_utils.erl]: Why the difference in naming for this function? Why not `find_document/1`?

File: src/els_indexer.erl [hunk 21]
```diff
@@ -67,20 +64,32 @@ index_file(Path, SyncAsync) ->
...
+%% TODO: Avoid nested transactions
+-spec index(els_dt_document:item()) -> ok.
+index(#{uri := Uri, text := Text} = Document) ->
+  MD5 = erlang:md5(Text),
+  case els_dt_document:lookup(Uri) of
+    {ok, [#{md5 := MD5}]} ->
+      %% The module is already indexed, no action needed.
+      ok;
+    _ ->
+      F = fun() ->
```
Comment (jfacorro) [inline src/els_indexer.erl]: For readability, would it be possible to move this out into it's own function that takes arguments?

File: src/els_dt_documents.erl (hunk 14 above)
Comment (jfacorro) [reply src/els_dt_documents.erl]: I don't have any better suggestions to be honest 😜

File: src/els_db.erl (hunk 5 above)
Comment (robertoaloi) [reply src/els_db.erl]: Already done in subsequent commits.

File: src/els_utils.erl (hunk 7 above)
Comment (robertoaloi) [reply src/els_utils.erl]: This was due to a bug in the db module. Fixed.

File: src/els_code_navigation.erl (hunk 15 above)
Comment (robertoaloi) [reply src/els_code_navigation.erl]: Fixed.

File: src/els_indexer.erl (hunk 21 above)
Comment (robertoaloi) [reply src/els_indexer.erl]: Yes, but I would leave this refactoring out of this PR. The indexer module needs some love.

File: rebar.config [hunk 22]
```diff
@@ -78,6 +78,7 @@
 {dialyzer, [ {warnings, [ unknown
                         ]}
            , {plt_apps, all_deps}
+           , {plt_extra_apps, [mnesia]}
```
Comment (robertoaloi) [inline rebar.config]: Due to the Mnesia setup (the application needs to be stopped when creating the schema), mnesia is not part of the erlang_ls `apps`. Therefore, Dialyzer would not include it as a dependency automatically, resulting in "unknown functions" warnings.

File: src/els_code_navigation.erl [hunk 23]
```diff
@@ -106,4 +101,6 @@ add_include_uri(#{ id := String }, Acc) ->
 beginning() ->
   #{range => #{from => {1, 1}, to => {1, 1}}}.
 
-%% TODO: Handle multiple header files with the same name?
+-spec filename_to_atom(string()) -> atom().
+filename_to_atom(FileName) ->
+  list_to_atom(filename:basename(FileName, filename:extension(FileName))).
```
Comment (robertoaloi) [inline src/els_code_navigation.erl]: Yes, dinamically creating atoms. But we can't always ensure they always exist (and therefore use the `to_existing_atom` variant here. Hopefully, this is not going to be a problem...

File: src/els_utils.erl (hunk 20 above)
Comment (robertoaloi) [reply src/els_utils.erl]: The idea was to distinguish between a lookup and a search operation.

**Comment by robertoaloi** [comment]
Merged because:

* I got an approval \o/
* it will be easier for us to test it on a daily basis
* the PR was huge and would conflict with almost every other PR
* CI did not show any regressions
* we can still incrementally fix/clean it
* The VSCode extension (which seems to be the most used way to interact with the server) still uses an older version

Feel free to add after-merge comments and I will address them.

**Pull request merged**